src/sgame/sg_active.cpp

```diff
@@ -1049,7 +1049,8 @@ void ClientTimerActions( gentity_t *ent, int msec )
 	{
 		if ( client->ps.ammo < BG_Weapon( WP_ALEVEL3_UPG )->maxAmmo )
 		{
-			if ( ent->timestamp + LEVEL3_BOUNCEBALL_REGEN < level.time )
+			int bounceballRegen = BG_GetBarbRegenerationInterval(ent->client->ps);
+			if ( ent->timestamp + bounceballRegen < level.time )
```

Comment:
The problem with this old approach is that I can shoot while not on creep or near a booster, wait in place for `ballRegenBooster` ms, then enter the booster range and instantly get a barb. This harms the intuition given by the pulse frequency. Alternatively, you could add a `gentity_t` float field where, each frame where you don't have full ammo, you add the frame time divided by the regen interval and when this value gets greater than 1, substract 1 from it (or reset it to 0) and add a barb. (Don't be afraid of adding single-purpose `gentity_t` fields, they will all go into components later.)